### DISCUSSION TITLE
Syntax error when trying to add function in Supabase user interface

### DISCUSSION THREAD

**Discussion by JoshuaeKaiser** [supabase/supabase#20277] [Questions]
I am able to add the function using PGAdmin with no issue, its runs and its fine. When I try to use the UI to add the function it will not let me save to what it calls syntax errors.

Been stuck on this for hours, and cannot figure it out.

This is the function im trying to add:

`create or replace function match_documents (
  query_embedding vector(1536),
  match_count int DEFAULT null,
  filter jsonb DEFAULT '{}'
) returns table (
  id bigint,
  content text,
  metadata jsonb,
  embedding jsonb,
  similarity float
)
language plpgsql
as $$
#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    (embedding::text)::jsonb as embedding,
    1 - (documents.embedding <=> query_embedding) as similarity
  from documents
  where metadata @> filter
  order by documents.embedding <=> query_embedding
  limit match_count;
end;
$$;
`

I have tried with every return type, still wont let me save. I have also tried starting at the begin and ommitting the arguments which I define in the user interface and it still will not work.

Somebody please help!

<img width="1792" alt="Screenshot 2024-01-09 at 13 00 56" src="https://github.com/supabase/supabase/assets/17850490/0c90ce37-b0bf-4c4f-b4f7-f7ba99471f1c">
<img width="1194" alt="Screenshot 2024-01-09 at 16 40 57" src="https://github.com/supabase/supabase/assets/17850490/91d1bad5-9a42-4ab7-894a-ecda636f9e01">

**Comment by GaryAustin1** [comment]
You only enter the actual function code in the function UI.  Not create function etc.  Normally the stuff starting with begin and ending with end for plpgsql.
You can also use the SQL editor in the dashboard to run the full SQL.
Note the UI cannot handle array, table or setof record returns.

EDIT: It does look like you want to return a table in your function code so that has to be done with the SQL editor or a 3rd party tool like you used.
